CodeModel/CodeParsers/CSharp/CaDETClassMetricCalculator.cs

```diff
@@ -37,10 +37,127 @@ internal class CaDETClassMetricCalculator
+            if (parsedClass.Parent == null) return 0;
+            return (double) CountProtectedMembersUsed(parsedClass) / CountProtectedMembersInBaseClass(parsedClass.Parent);
+        }
+
+        private static int CountProtectedMembersInBaseClass(CaDETClass parent)
+        {
+            return FindFieldsWithModifier(parent.Fields, CaDETModifierValue.Protected).Count() + FindMembersWithModifier(parent.Members, CaDETModifierValue.Protected).Count();
+        }
+
+        private static int CountProtectedMembersUsed(CaDETClass parsedClass)
+        {
+            return CountAccessedInheritatedFields(parsedClass) + CountUsedInheritatedMethods(parsedClass);
+        }
+
+        private static int CountAccessedInheritatedFields(CaDETClass parsedClass)
+        {
+            var protectedParentFields = FindFieldsWithModifier(parsedClass.Parent.Fields, CaDETModifierValue.Protected);
+            var accessedFields = parsedClass.Members.SelectMany(m => m.AccessedFields);
+            var accessedInheritedFields = protectedParentFields.Select(f => f.Name).Intersect(accessedFields.Select(f => f.Name));
+            return accessedInheritedFields.Count();
+        }
+
+        private static int CountUsedInheritatedMethods(CaDETClass parsedClass)
+        {
+            var protectedParentMembers = FindMembersWithModifier(parsedClass.Parent.Members, CaDETModifierValue.Protected);
+            var invokedMethods = parsedClass.Members.SelectMany(m => m.InvokedMethods);
+            var accessedAccessors = parsedClass.Members.SelectMany(m => m.AccessedAccessors);
+            var usedMethods = invokedMethods.Union(accessedAccessors);
+            var usedInheritedMethods = protectedParentMembers.Select(m => m.Name).Intersect(usedMethods.Select(m => m.Name));
+            return usedInheritedMethods.Count();
+        }
+
+        private static double GetWMCOfNotAccessorOrMuttatorMethods(CaDETClass parsedClass)
+        {
+            return parsedClass.Members.Where(m => !m.Type.Equals(CaDETMemberType.Property)).Sum(m => m.Metrics[CaDETMetric.CYCLO]);
+        }
+
+        // Public attributes for NOPA metric do not include constants and static fields. (Object-oriented metrics in practice)
+        private static int CountPublicAttributes(CaDETClass parsedClass)
+        {
+            var publicFields = GetPublicFields(parsedClass.Fields);
+            var staticFields = (from field in publicFields
+                             from modifier in field.Modifiers
+                             where modifier.Value.Equals(CaDETModifierValue.Static)
+                             select field).ToList();
+            return publicFields.Where(f => !staticFields.Contains(f)).Count();
+        }
+
+        private static double CountWeightOfClass(CaDETClass parsedClass)
+        {
+            return (double)CountFunctionalPublicMethods(parsedClass.Members) / (CountPublicProperties(parsedClass.Members) + GetPublicFields(parsedClass.Fields).Count());
```

Comment:
You should try to keep the ordering of the private functions to follow the expression. For this method, I would put CountFunctionalPublicMethods, then CountPublicProperties, and then GetPublicFields. It's easier to read that way.